README.md
## Intergrid: interpolate data given on an N-d rectangular grid

Purpose: interpolate data given on an N-d rectangular grid,
uniform or non-uniform,
using the fast
[scipy.ndimage.map_coordinates](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.interpolation.map_coordinates.html).
Non-uniform grids are first uniformized with
[numpy.interp](https://docs.scipy.org/doc/numpy/reference/generated/numpy.interp.html).

Keywords, tags: interpolation, rectangular grid, box grid, python, numpy, scipy

Background:
the reader should know some Python and NumPy
([IPython](https://ipython.org) is invaluable for learning both).
For basics of interpolation, see
[Bilinear interpolation](https://en.wikipedia.org/wiki/Bilinear_interpolation)
on Wikipedia. For `map_coordinates`, see the example under
[multivariate-spline-interpolation-in-python-scipy](https://stackoverflow.com/questions/6238250/multivariate-spline-interpolation-in-python-scipy)
on stackoverflow.

### Example

Say we have rainfall on a 4 x 5 grid of rectangles, lat 52 .. 55 x lon -10 .. -6,
and want to interpolate (estimate) rainfall at 1000 query points
in between the grid points.

	from intergrid.intergrid import Intergrid  # .../intergrid/intergrid.py

        # define the grid --
    griddata = np.loadtxt(...)  # griddata.shape == (4, 5)
    lo = np.array([ 52, -10 ])  # lowest lat, lowest lon
    hi = np.array([ 55, -6 ])   # highest lat, highest lon

        # set up an interpolator function "interfunc()" with class Intergrid --
    interfunc = Intergrid( griddata, lo=lo, hi=hi )

        # generate 1000 random query points, lo <= [lat, lon] <= hi --
    query_points = lo + np.random.uniform( size=(1000, 2) ) * (hi - lo)

        # get rainfall at the 1000 query points --
    query_values = interfunc.at( query_points )  # -> 1000 values

What this does: for each [lat, lon] in query_points,

1) find the square of `griddata` it's in,
e.g. [52.5, -8.1] -> [0, 3] [0, 4] [1, 4] [1, 3]\
2) do bilinear (multilinear) interpolation in that square,
using `scipy.ndimage.map_coordinates` .

Check:\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; `interfunc( lo ) == griddata[0, 0]`\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; `interfunc( hi ) == griddata[-1, -1]` i.e. `griddata[3, 4]`


### Parameters

`griddata`: numpy array_like, 2d 3d 4d ...\
`lo, hi`: user coordinates of the corners of griddata, 1d array-like, lo < hi\
`maps`: an optional list of `dim` descriptors of piecewise-linear or nonlinear maps,\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; e.g. [[50, 52, 62, 63], None] \ \ # uniformize lat, linear lon; see below\
`copy`: make a copy of query_points, default `True`;\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; `copy=False` overwrites query_points, runs in less memory\
`verbose`: the default 1 prints a summary of each call, with run time\
`order`: interpolation order:\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; default 1: bilinear, trilinear ... interpolation using all 2^dim corners\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; 0: each query point -> the nearest grid point -> the value there\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; 2 to 5: spline, see below\
`prefilter`: the kind of spline:\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; default `False`: smoothing B-spline\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; `True`: exact-fit C-R spline\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; 1/3: Mitchell-Netravali spline, 1/3 B + 2/3 fit


### Methods

After setting up `interfunc = Intergrid(...)`, either

    query_values = interfunc.at( query_points )  # or
    query_values = interfunc( query_points )

do the interpolation. (The latter is `__call__` in python.)


### Non-uniform rectangular grids

What if our griddata above is at non-uniformly-spaced latitudes,
say [50, 52, 62, 63] ?  `Intergrid` can "uniformize" these
before interpolation, like this:

    lo = np.array([ 50, -10 ])
    hi = np.array([ 63, -6 ])
    maps = [[50, 52, 62, 63], None]  # uniformize lat, linear lon
    interfunc = Intergrid( griddata, lo=lo, hi=hi, maps=maps )

This will map (transform, stretch, warp) the lats in query_points column 0
to array coordinates in the range 0 .. 3, using `np.interp` to do
piecewise-linear (PWL) mapping:

    50  51  52  53  54  55  56  57  58  59  60  61  62  63   # lo[0] .. hi[0]
     0  .5   1  1.1 1.2 1.3 1.4 1.5 1.6 1.7 1.8 1.9  2   3

`maps[1] None` says to map the lons in query_points column 1 linearly:

    -10  -9  -8  -7  -6   # lo[1] .. hi[1]
      0   1   2   3   4

Mapping details

The query_points are first clipped, then columns mapped linearly or non-linearly,
then fed to `map_coordinates` .\
In `dim` dimensions (i.e. axes or columns), `lo` and `hi` are each `dim` numbers,
the low and high corners of the data grid.\
`maps` is an optional list of `dim` map descriptors, which can be

* `None`: linear-transform that column, `query_points[:,j]`, to `griddata`:\
    `lo[j] -> 0`\
    `hi[j] -> griddata.shape[j] - 1`
* a callable function: e.g. `np.log` does\
    `query_points[:,j] = np.log( query_points[:,j] )`
* a *sorted* array describing a non-uniform grid:\
    `query_points[:,j] =`\
    `np.interp( query_points[:,j], [50, 52, 62, 63], [0, 1, 2, 3] )`


### Download

    git clone https://github.com/denis-bz/intergrid.git
        # ? pip install --user git+https://github.com/denis-bz/intergrid.git
        # ? pip install --user intergrid

    # tell python where the intergrid directory is, e.g. in your ~/.bashrc:
    #   export PYTHONPATH=$PYTHONPATH:.../intergrid/

    # test in python or IPython:
    from intergrid.intergrid import Intergrid  # i.e. .../intergrid/intergrid.py


### Splines

`Intergrid( ... order = 0 to 5 )` gives the spline order:\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; `order=1`, the default, does bilinear, trilinear ...
interpolation, which looks at the grid data at all 4 8 16 .. 2^dim corners of
the box around each query point.\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; `order=0` looks at only the one gridpoint
nearest each query point &mdash; crude but fast.\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; `order = 2 to 5` does spline interpolation on a uniform
or uniformized grid, looking at (order+1)^dim neighbors of each query point.

`Intergrid( ... prefilter = False | True | 1/3 )`
specifies the kind of spline, for `order >= 2`:\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; `prefilter=0` or `False`, the default: B-spline\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; `prefilter=1` or `True`: exact-fit spline\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; `prefilter=1/3`: M-N spline.\
A B-spline goes through smoothed data points,
with [1 4 1] smoothing, [0 0 1 0 0] -> [0 1 4 1 0] / 6.\
An exact-fit a.k.a interpolating spline
goes through the data points exactly.
This is not what you want for noisy data,
and may also wiggle or overshoot more than B-splines do.\
An M-N spline blends 1/3 B-spline and 2/3 exact-fit; see
Mitchell and Netravali,
[Reconstruction filters in computer-graphics](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.125.201&rep=rep1&type=pdf) ,
1988, and the plots from `intergrid/test/MNspline.py`.

<small>
Fine print: Exact-fit or interpolating splines can be local or global.
Catmull-Rom splines and the original M-N splines are local:
they look at 4 neighbors of each query point in 1d, 16 in 2d, 64 in 3d.
Prefiltering is global, with IIR falloff ~ 1 / 4^distance.
(I don't know of test images that show a visible difference to local C-R).
Confusingly, the term "Cardinal spline" is sometimes used
for local (C-R, FIR),
and sometimes for global (IIR prefilter, then B-spline).

Prefiltering is a clever transformation
such that `Bspline( transform( data )) = exactfitspline( data )`.
It is described in a paper by M. Unser,
[Splines: A perfect fit for signal and image processing](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.19.6706&rep=rep1&type=pdf) ,
1999.
</small>

Uniformizing a grid with PWL, then uniform-splining, is fast and simple,
but not as smooth as true splining on the original non-uniform grid.
The differences will of course depend on the grid spacings
and on how rough the function is.


### Notes

Run any interpolator on *your* data with orders 0, 1 ...
to get an idea of how the results get smoother, and take longer.
Check a few query points by hand; plot some cross-sections.

`griddata` values can be of any numpy integer or floating type: int8 uint8
.. int32 int64 float32 float64.
Beware of overflow: interpolating uint8 s can give values outside the range 0 .. 255.
(Interpolation in `d` dimensions can overshoot by (9/8)^d .)
`np.float32` will use less memory than `np.float64`,
but beware of functions in the flow that silently convert everything
to float64.  The values must be numbers, not vectors.

Coordinate scaling doesn't matter to `Intergrid`;
corner weights are calculated in unit cubes of `griddata`,
after scaling and mapping. If for example griddata column 3
is multiplied by 1000, and lo[3] hi[3] too, the weights are unchanged.

Box grids get big and slow above 5d.
A cube with steps 0 .1 .2 .. 1.0 in all dimensions
has 11^6 ~ 1.8M points in 6d, 11^8 ~ 200M in 8d.
One can reduce that only with a coarser grid like 0 .5 1 in some dimensions
(those that vary the least).
But time ~ 2^d per query point grows pretty fast.

`map_coordinates` in 5d with `order=1` looks at 32 corner values, with average weight 3 %.
If the weights are roughly equal
(which they will tend to be, by the central limit theorem ?),
sharp edges or gradients will be blurred, and colors mixed to a grey fog.

To see how different interpolators affect images, run matplotlib
`plt.imshow( interpolation = "nearest" / "bilinear" / ... )` .


### Kinds of grids

Terminology varies, so the basic kinds of box grids
a.k.a. rectangular grids are defined here.

An integer or Cartesian grid has integer coordinates,
e.g.  2 x 3 x 5 points in a numpy array:
`A = np.array((2,3,5)); A[0,0,0], A[0,0,1] .. A[1,2,4]`.

A uniform box grid has nx x ny x nz ... points uniformly spaced,
linspace x linspace x linspace ...
so all boxes have the same size and are axis-aligned.
Examples: 1024 x 768 pixels on a screen,
or 4 x 5 points at latitudes [10 20 30 40] x longitudes [-10 -9 -8 -7 -6].

A non-uniform box grid also has nx x ny x nz ... points,
but allows non-uniform spacings,
e.g. latitudes [-10 0 60 70] x longitudes [-10 -9 0 20 40];
the boxes have different sizes but are still axis-aligned.

(Scattered data, as the name says, has points anywhere,
not only on grid lines.
To interpolate scattered data in `scipy`, see
[scipy.interpolate.griddata](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.griddata.html)
and
[scipy.spatial.cKDTree](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.cKDTree.html)
.)

There are countless varieties of grids:
grids with holes, grids warped to various map projections,
multiscale / multiresolution grids ...


### Run times

See intergrid/test/test-4d.py: a 4d grid with 1M scattered query points,
uniform / non-uniform box grid, on a 2.5Gz i5 iMac:

    shape (361, 720, 47, 8)  98M * 8
    Intergrid: 617 msec  1000000 points in a (361, 720, 47, 8) grid  0 maps  order 1
    Intergrid: 788 msec  1000000 points in a (361, 720, 47, 8) grid  4 maps  order 1


### See also

[scipy.ndimage.interpolation.map_coordinates](https://docs.scipy.org/doc/scipy-dev/reference/generated/scipy.ndimage.interpolation.map_coordinates.html)\
[scipy reference ndimage](https://docs.scipy.org/doc/scipy/reference/tutorial/ndimage.html)\
[stackoverflow.com/questions/tagged/scipy+interpolation](https://stackoverflow.com/questions/tagged/scipy+interpolation)\
[interpol 2014](https://github.com/denis-bz/interpol) -- intergrid + barypol\
Google "regrid | resample"\
`pip search interpol` (also gets string interpolation)


### Comments welcome

and testcases most welcome\
    &mdash; denis-bz-py at t-online dot de



changelog.txt
## 2013-06-04 Jun  prefilter Bspline / exact-fit / M-N
    https://github.com/denis-bz/interpol.git is unchanged:
    intergrid/* barypol/* and test/* .

## 2020-02-20 Feb
    https://github.com/denis-bz/intergrid.git: intergrid/* and test/*, dropped barypol.
    Minor mods for Python3.
	BSD-3 license
    (I'm trying to get `pypi install [--user / --prefix <dir>] intergrid` to work,
	no fun at all.)



intergrid/test/util.py
# intergrid/test/util.py

from __future__ import division
import numpy as np

def avmaxdiff( interpol, exact, query_points=None ):
    absdiff = np.fabs( interpol - exact )
    av = absdiff.mean()
    jmax = absdiff.argmax()  # flat
    at = "at %s" % query_points[jmax]  if query_points is not None \
        else ""
    return "av %.2g  max %.2g = %.3g - %.3g  %s" % (
            av, absdiff[jmax], interpol[jmax], exact[jmax], at )



intergrid/test/testfuncs.py
#!/usr/bin/env python
# testfuncs.py: [ f( X npt x dim in unit cube ) ... ]
# scaled so all max grad ~ 1  for dim 4 side 20

from __future__ import division, print_function
import numpy as np
from numpy import pi, cos, sin

np.set_printoptions( threshold=20, edgeitems=10, linewidth=140,
        formatter = dict( float = lambda x: "%.2g" % x ))  # float arrays %.2g

#...............................................................................
def cos_xav( X, freq=1, **kw ):
    """ cos 2 pi f Xav, X in unit cube
        Xav ~ sqrt(dim) / 2
        freq 1 dim 4:
            axis 1 .. 0   cos 2 pi xav = x / dim
            diag 1 .. -1 .. 1   cos 2 pi xav = x
    """
    return cos( 2 * pi * freq * X.mean(axis=1) ) / .004

def volripple( X, fm=6, alpha=.25, **kw ):
    """ npt x dim -> npt
    from Marschner & Lobb, reconstruction filters for volume rendering
    side 20 / h .05 in 3d:
    """
    X = np.asanyarray(X)
    Xdim = X.ndim
    if Xdim == 1:
        X = np.array([X])
    r = np.sqrt( (X[:,:-1] ** 2) .sum(axis=1) )
    z = X[:,-1]
    coscos = cos( 2*pi*fm * cos( pi * r / 2 ))
    rho = (1 - sin( pi * z / 2 ) + alpha * (1 + coscos)) \
        / (2 * (1 + alpha))
    return (rho if Xdim == 2  else rho[0]) / .01

def linear_should_be_exact( X ):
    return X.mean(axis=1)

def parabola( X ):
    xav = X.mean(axis=1)
    return (xav**2 - xav) / .0006

def sawtooth( X, c=.1 ):
    return (c * X.sum(axis=1)) % 1


testfuncs = [cos_xav, volripple, parabola]

#...............................................................................
if __name__ == "__main__":
    import sys

    dim = 3
    side = 20  # grid side^dim

    # to change these params, run this.py  a=1  b=None  'c = expr' ... in sh or ipython --
    for arg in sys.argv[1:]:
        exec( arg )

    side |= 1
    shape = dim * [side]
    grid = np.array( list( np.ndindex(*shape) ), float ) / (side - 1)

    for func in testfuncs:
        print("\n%s --" % func.__name__)
        funcgrid = func( grid ).reshape(shape)
        # print funcgrid
        absgrad = np.fabs( np.gradient( funcgrid, *shape ))  # dim, shape
        jmax = absgrad.argmax()
        # at = np.unravel_index( jmax, absgrad.shape )
        print("max grad: %.2g at %s" % (
            absgrad.ravel()[jmax], grid[ jmax // dim ]))



intergrid/test/test-intergrid.py
#!/usr/bin/env python
# test-interpol.py dim= side= ...
# for testfunc in [...]:
# for method in [Intergrid ...]:
#   |interpolated - exact| uniform-random query pointsin side^dim

from __future__ import division, print_function
import sys
from time import time
import numpy as np
import scipy

from intergrid.intergrid import Intergrid  # .../intergrid/intergrid.py
from testfuncs import testfuncs  # [ cos_xav ... on unit cube ]
import util as ut

np.set_printoptions( threshold=20, edgeitems=10, linewidth=140,
        formatter = dict( float = lambda x: "%.2g" % x ))  # float arrays %.2g

print(80 * "=")
print("versions: numpy %s  scipy %s  python %s" % (
    np.__version__, scipy.__version__, sys.version.split()[0] ))

#...............................................................................
methods = [Intergrid]
dim = 3
side = 20  # uniform grid side^dim
nquery = 1e6
jmethod = -1
jfunc = -1
dtype = np.float64  # map_coord only
orders = [1, 2]
save = 0
seed = 0

# to change these params, run this.py  a=1  b=None  'c = expr' ... in sh or ipython --
for arg in sys.argv[1:]:
    exec( arg )

np.random.seed(seed)

nquery = int(nquery)
if jmethod >= 0:
    methods = [methods[jmethod]]
if jfunc >= 0:
    testfuncs = [testfuncs[jfunc]]
shape = dim * [side]

    # side^d x d  000 .. 111 --
grid = np.array( list( np.ndindex(*shape) ), dtype ) / (side - 1)
lo = np.array( dim * [0.] )
hi = np.array( dim * [1.] )
query = np.random.uniform( size=(nquery, dim) ).astype(dtype)

#...............................................................................
params = "%s  shape %s  nquery %d" % (__file__, shape, nquery)
print( params )

for testfunc in testfuncs:
    funcname = testfunc.__name__
    funcgrid = testfunc( grid ).reshape(shape)
    print("")
    exact = testfunc(query)

    for method in methods:
        methodname = method.__name__
        query_out = np.empty( nquery )

        for order in orders:
            interpolator = method( funcgrid, lo=lo, hi=hi, order=order, verbose=0 )
            t0 = time()  # wallclock

            interpolator.at( query, query_out )
            ms = (time() - t0) * 1000

            diffs = ut.avmaxdiff( query_out, exact, query )
            print( "%4d ms  |%s - exact|  %s  order %d: %s " % (
                    ms, methodname, funcname, order, diffs ))

            if save:  # to plot
                sav = "%dd-order%d-%s-%s.npz" % (dim, order, methodname, funcname)
                print( "np.savez >", sav )
                np.savez( sav, query_out=query_out.astype(np.float32),
                        diffs=diffs, params=params )



intergrid/intergrid.py
#!/usr/bin/env python
""" interpolate data given on an Nd rectangular grid, uniform or non-uniform.

Purpose: extend the fast N-dimensional interpolator
`scipy.ndimage.map_coordinates` to non-uniform grids, using `np.interp`.

Background: please look at
http://en.wikipedia.org/wiki/Bilinear_interpolation
http://stackoverflow.com/questions/6238250/multivariate-spline-interpolation-in-python-scipy
http://docs.scipy.org/doc/scipy-dev/reference/generated/scipy.ndimage.interpolation.map_coordinates.html

Example
-------
Say we have rainfall on a 4 x 5 grid of rectangles, lat 52 .. 55 x lon -10 .. -6,
and want to interpolate (estimate) rainfall at 1000 query points
in between the grid points.

    from intergrid.intergrid import Intergrid  # .../intergrid/intergrid.py

        # define the grid --
    griddata = np.loadtxt(...)  # griddata.shape == (4, 5)
    lo = np.array([ 52, -10 ])  # lowest lat, lowest lon
    hi = np.array([ 55, -6 ])   # highest lat, highest lon

        # set up an interpolator function "interfunc()" with class Intergrid --
    interfunc = Intergrid( griddata, lo=lo, hi=hi )

        # generate 1000 random query points, lo <= [lat, lon] <= hi --
    query_points = lo + np.random.uniform( size=(1000, 2) ) * (hi - lo)

        # get rainfall at the 1000 query points --
    query_values = interfunc( query_points )  # -> 1000 values

What this does:
    for each [lat, lon] in query_points:
        1) find the square of griddata it's in,
            e.g. [52.5, -8.1] -> [0, 3] [0, 4] [1, 4] [1, 3]
        2) do bilinear (multilinear) interpolation in that square,
            using `scipy.ndimage.map_coordinates` .
Check:
    interfunc( lo ) -> griddata[0, 0],
    interfunc( hi ) -> griddata[-1, -1] i.e. griddata[3, 4]

Parameters
----------
    griddata: numpy array_like, 2d 3d 4d ...
    lo, hi: user coordinates of the corners of griddata, 1d array-like, lo < hi
    maps: a list of `dim` descriptors of piecewise-linear or nonlinear maps,
        e.g. [[50, 52, 62, 63], None]  # uniformize lat, linear lon
    copy: make a copy of query_points, default True;
        copy=False overwrites query_points, runs in less memory
    verbose: default 1: print( a 1-line summary for each call, with run time
    order=1: see `map_coordinates`
    prefilter: 0 or False, the default: smoothing B-spline
              1 or True: exact-fit interpolating spline (IIR, not C-R)
              1/3: Mitchell-Netravali spline, 1/3 B + 2/3 fit
        (prefilter is only for order > 1, since order = 1 interpolates)

Non-uniform rectangular grids
-----------------------------
What if our griddata above is at non-uniformly-spaced latitudes,
say [50, 52, 62, 63] ?  `Intergrid` can "uniformize" these
before interpolation, like this:

    lo = np.array([ 50, -10 ])
    hi = np.array([ 63, -6 ])
    maps = [[50, 52, 62, 63], None]  # uniformize lat, linear lon
    interfunc = Intergrid( griddata, lo=lo, hi=hi, maps=maps )

This will map (transform, stretch, warp) the lats in query_points column 0
to array coordinates in the range 0 .. 3, using `np.interp` to do
piecewise-linear (PWL) mapping:
    50  51  52  53  54  55  56  57  58  59  60  61  62  63  # lo[0] .. hi[0]
    0   .5  1   1.1 1.2 1.3 1.4 1.5 1.6 1.7 1.8 1.9 2   3

`maps[1] None` says to map the lons in query_points column 1 linearly:
    -10  -9  -8  -7  -6  # lo[1] .. hi[1]
    0    1   2   3   4

More doc: https://github.com/denis-bz/intergrid/blob/master/README.md

"""
# split class Gridmap ?

from __future__ import division, print_function
from time import time
# warnings
import numpy as np
from scipy.ndimage import map_coordinates, spline_filter

__author_email__ = "denis-bz-py@t-online.de"  # comments welcome, testcases most welcome

#...............................................................................
class Intergrid:
    __doc__ = globals()["__doc__"]

    def __init__( self, griddata, lo, hi, maps=None, copy=True, verbose=1,
            order=1, prefilter=False ):
        griddata = np.asanyarray( griddata )
        dim = griddata.ndim  # - (griddata.shape[-1] == 1)  # ??
        assert dim >= 2, griddata.shape
        self.dim = dim
        if np.isscalar(lo):
            lo *= np.ones(dim)
        if np.isscalar(hi):
            hi *= np.ones(dim)
        self.loclip = lo = np.asarray_chkfinite( lo ).copy()
        self.hiclip = hi = np.asarray_chkfinite( hi ).copy()
        assert lo.shape == (dim,), lo.shape
        assert hi.shape == (dim,), hi.shape
        self.copy = copy
        self.verbose = verbose
        self.order = order
        if order > 1  and 0 < prefilter < 1:  # 1/3: Mitchell-Netravali = 1/3 B + 2/3 fit
            exactfit = spline_filter( griddata )  # see Unser
            griddata += prefilter * (exactfit - griddata)
            prefilter = False
        self.griddata = griddata
        self.prefilter = (prefilter == True)

        if maps is None:
            maps = [None,] * len(lo)

        self.maps = maps
        self.nmap = 0
        if len(maps) > 0:
            assert len(maps) == dim, "maps must have len %d, not %d" % (
                    dim, len(maps))
            # linear maps (map None): Xcol -= lo *= scale -> [0, n-1]
            # nonlinear: np.interp e.g. [50 52 62 63] -> [0 1 2 3]
            self._lo = np.zeros(dim)
            self._scale = np.ones(dim)

            for j, (map, n, l, h) in enumerate( zip( maps, griddata.shape, lo, hi )):
                ## print( "test: j map n l h:", j, map, n, l, h )
                if map is None  or callable(map):
                    self._lo[j] = l
                    if h > l:
                        self._scale[j] = (n - 1) / (h - l)  # _map lo -> 0, hi -> n - 1
                    else:
                        self._scale[j] = 0  # h <= l: X[:,j] -> 0
                    continue
                self.maps[j] = map = np.asanyarray(map)
                self.nmap += 1
                assert len(map) == n, "maps[%d] must have len %d, not %d" % (
                    j, n, len(map) )
                mlo, mhi = map.min(), map.max()
                if not (l <= mlo <= mhi <= h):
                    print( "Warning: Intergrid maps[%d] min %.3g max %.3g " \
                        "are outside lo %.3g hi %.3g" % (
                        j, mlo, mhi, l, h ))

#...............................................................................
    def _map_to_uniform_grid( self, X ):
        """ clip, map X linear / nonlinear  inplace """
        np.clip( X, self.loclip, self.hiclip, out=X )
            # X nonlinear maps inplace --
        for j, map in enumerate(self.maps):
            if map is None:
                continue
            if callable(map):
                X[:,j] = map( X[:,j] )  # clip again ?
            else:
                    # PWL e.g. [50 52 62 63] -> [0 1 2 3] --
                X[:,j] = np.interp( X[:,j], map, np.arange(len(map)) )

            # linear map the rest, inplace (nonlinear _lo 0, _scale 1: noop)
        if self.nmap < self.dim:
            X -= self._lo
            X *= self._scale  # (griddata.shape - 1) / (hi - lo)
        ## print( "test: _map_to_uniform_grid", X.T )

#...............................................................................
    def __call__( self, X, out=None ):
        """ query_values = Intergrid(...) ( query_points npt x dim )
        """
        X = np.asanyarray(X)
        assert X.shape[-1] == self.dim, ("the query array must have %d columns, "
                "but its shape is %s" % (self.dim, X.shape) )
        Xdim = X.ndim
        if Xdim == 1:
            X = np.asarray([X])  # in a single point -> out scalar
        if self.copy:
            X = X.copy()
        assert X.ndim == 2, X.shape
        npt = X.shape[0]
        if out is None:
            out = np.empty( npt, dtype=self.griddata.dtype )
        t0 = time()
        self._map_to_uniform_grid( X )  # X inplace
#...............................................................................
        map_coordinates( self.griddata, X.T,
            order=self.order, prefilter=self.prefilter,
            mode="nearest",  # outside -> edge
                # test: mode="constant", cval=np.NaN,
            output=out )
        if self.verbose:
            print( "Intergrid: %.3g msec  %d points in a %s grid  %d maps  order %d" % (
                (time() - t0) * 1000, npt, self.griddata.shape, self.nmap, self.order ))
        return out if Xdim == 2  else out[0]

    at = __call__

# end intergrid.py

